Validation for active infractions

SebastiaanZ opened the pull request:
This pull request adds validation rules to the API to validate infractions based on the `active` field. This means that the API will now reject active infractions of types that should never be active (notes, warnings, and kicks) and only accept one active infraction per user for the other types of infractions (mute, ban, watch, superstar). 

In addition, this pull requests adds a data migration to make sure the database itself is consistent with the new validation rules. Since this is a fairly complex data migration, I have added tests to make sure the migration works as expected. (Read more about that below.)

Finally, I've introduced a database-level `UniqueConstraint` to prevent a user from having multiple active infractions of the same type. I have also added tests to make sure this constraint works as expected.

## Testing migrations

To test the effect of the migration file, I've created a subclass of `django.test.TestCase`, `MigrationsTestCase`, that reverts the test database to the migration just prior to the one of interest. We can then inject test data, apply the migration, and test if the migration had the desired effect. 

Since a migration is really snapshot in time of the migration history, this also allows us to work with the database models as they were at those two time points in the migration history (directly before and directly after applying this migration), since otherwise these tests will fail in the future if we decide to change the models.

SebastiaanZ committed:
Add validation rules to Infraction serializer

https://github.com/python-discord/site/issues/273

This commit adds validation rules to the Infraction serializer that
validate if a given infraction should be accepted based on its status
of being considered `active`. If the validation fails, the API will
reject the request and return a 400 status.

Specifically, this validator checks that:

- infractions that can never be active do not have `active=True` set;

- a user can never receive a second active infraction of the same type.

Tests have been added to `test_infractions.py` to ensure that the
validators work as expected.

This commit implements the first part of #273

SebastiaanZ committed:
Migrate undesirable active infraction to inactive

https://github.com/python-discord/site/issues/273

This commit adds a data migration to migrate active infractions that
should not be active to inactive. There are two types of infractions
that this migration will migrate to inactive:

- Infractions of types that should never be active (e.g. notes)

- Secondary active infractions if a given user already has an active
  infraction of the same type.

Since this makes the migration file fairly complex, I have written
tests to make sure the migration works as expected. In order to do
this, I've subclassed `django.test.TestCase` to create a
`MigrationsTestCase` that takes care of reverting the database back
to a state prior to the migrations we want to test and injects test
data before applying the migrations we want to test.

For more information, see `pydis_site.apps.api.tests.migrations.base`

This implements the last part of and closes #273

SebastiaanZ committed:
Prevent double active infractions with constraint

https://github.com/python-discord/site/issues/273

This commits adds a UniqueConstraint for active infractions on a
combination of the `user` and `type` field. This means that a user
can only have one active infraction of a given type in the database
at any time.

I've also added tests to make sure that this behaves as expected.

SebastiaanZ commented on pydis_site/apps/api/serializers.py:
```diff
@@ -1,4 +1,5 @@
 """Converters from Django models to data interchange formats and back."""
+import logging
```
I've accidentally left this logging import here; I'll remove it when pushing the commit that straightens out the migration conflict that #269 will cause.

SebastiaanZ commented on pydis_site/apps/api/serializers.py:
```diff
@@ -12,6 +13,8 @@
     Tag, User
 )
 
+log = logging.getLogger(__name__)
```
I'll remove this then as well, obviously.

MarkKoz commented on pydis_site/apps/api/models/bot/user.py:
```diff
@@ -50,7 +50,7 @@ class User(ModelReprMixin, models.Model):
     def __str__(self):
         """Returns the name and discriminator for the current user, for display purposes."""
-        return f"{self.name}#{self.discriminator}"
+        return f"{self.name}#{self.discriminator:0>4}"
```
What's this for?

Edit: it's for when discriminators are less than 4 digits, right? Cause they are stored as integers and are therefore not zero-padded by default?

SebastiaanZ replied:
That's right, yeah.

MarkKoz commented on pydis_site/apps/api/tests/test_infractions.py:
```diff
@@ -305,6 +312,167 @@ def test_returns_400_for_non_hidden_required_hidden_type(self):
+                    'reason': 'Take me on!',
+                    'hidden': True,
+                    'active': True,
+                    'expires_at': '2019-10-04T12:52:00+00:00'
```
This test is a bit fragile in the sense that it relies on the active validation logic preceding the validation for allowing expiration or being hidden. But I think I can live with that.

SebastiaanZ replied:
I can make the payload more dependent on the type, but it would mean a logically more complex test.

MarkKoz replied:
Maybe not that much more complex. I'm wondering how it'd look if the expiration and hidden args could be put in the tuple alongside the types.

SebastiaanZ replied:
Better, I think.

MarkKoz commented on pydis_site/apps/api/serializers.py:
```diff
@@ -110,6 +113,16 @@ def validate(self, attrs: dict) -> dict:
+            raise ValidationError({'active': [f'{infr_type} infractions cannot be active.']})
+
+        user = attrs.get('user')
+        if active and Infraction.objects.filter(user=user, type=infr_type, active=True).exists():
```
How come a [`UniqueValidator`](https://www.django-rest-framework.org/api-guide/validators/#uniquevalidator) wasn't used instead? Also, supposedly a `ModelSerializer` does this: 

> It will automatically generate validators for the serializer, such as unique_together validators.

I wonder, does it support generating validators for unique constraints on the model?

SebastiaanZ replied:
I've changed this logic for the `UniqueTogetherValidator` with a custom message. It works well, although it gives slightly less specific feedback:

```py
'non_field_errors': ['This user already has an active infraction of this type.']
```

However, since this is generated for a specific infraction payload, that should not really be an issue and the message is unambiguous in the context of the query.

MarkKoz replied:
So you still ended up having to manually specify the validator. It did not generate it on its own?

SebastiaanZ replied:
As far as I can tell, it does not automatically add a validator for the constraint we've added to the model. If I remove our "custom" validator, the serializer doesn't seem to catch double active infractions and we'll get an `IntegrityError` (fail late in the database) instead.

MarkKoz replied:
That's fine then. I looked at the [source](https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L1535) and it seems to only work if [`unique_together`](https://docs.djangoproject.com/en/2.2/ref/models/options/#unique-together) is used.

MarkKoz commented on pydis_site/apps/api/tests/test_infractions.py:
```diff
@@ -305,6 +312,167 @@ def test_returns_400_for_non_hidden_required_hidden_type(self):
+                reason="The second active ban"
+            )
+
+    def test_unique_constraint_accepts_active_infraction_after_inactive_infraction(self):
```
Since this and the following test cases have no assertions, it means you're relying on `IntegrityError` not being raised? If it was raised (like any other exception), it'd fail the test, right? Sorry if that has an obvious answer, just making sure here.

SebastiaanZ replied:
I will change this to a more explicit variant. I've looked around for how to properly test a `assertNotRaises` situation, since there's no built-in way to do that with `unittest`, but I'm finding conflicting answers.

Some advise to just run it, like I've done here, and rely on the test resulting in `Error` if an exception occurs, but other argue, and I think I agree with that, that observing an exception where non should occur should result in a test `Failure` instead. 

One solution would be to do something like:

```py
try:
    # the model creation
except IntegrityError:
    self.fail("Raised IntegrityError unexpectedly!")
```

SebastiaanZ replied:
I've changed it to the proposed solution.

MarkKoz commented on pydis_site/apps/api/tests/test_infractions.py:
```diff
@@ -305,6 +312,167 @@ def test_returns_400_for_non_hidden_required_hidden_type(self):
+                }
+                response = self.client.post(url, data=invalid_infraction)
+                self.assertEqual(response.status_code, 400)
+                self.assertEqual(response.json(), {
```
I don't believe this is consistent with our code style but honestly it currently looks better than what it's "supposed" to be:

```py
self.assertEqual(
    response.json(), 
    {
        'active': [f'{infraction_type} infractions cannot be active.']
    }
)
```

It wouldn't be as bad if the dict was on one line:

```py
self.assertEqual(
    response.json(), 
    {'active': [f'{infraction_type} infractions cannot be active.']}
)
```

MarkKoz replied:
Oh, so I'm not sure if this should be changed or not. Thoughts?

SebastiaanZ replied:
I'll change that.

MarkKoz commented on pydis_site/apps/api/tests/test_infractions.py:
```diff
@@ -305,6 +312,167 @@ def test_returns_400_for_non_hidden_required_hidden_type(self):
+                second_response = self.client.post(url, data=second_active_infraction)
+                self.assertEqual(second_response.status_code, 400)
+                self.assertEqual(second_response.json(), {
+                    'active': [f'This user already has an active {infraction_type} infraction']
```
Same as [line 333](https://github.com/python-discord/site/pull/278/files#r333130421).

MarkKoz commented on pydis_site/apps/api/migrations/0044_active_infractions_migration.py:
```diff
@@ -0,0 +1,106 @@
+
+def reverse_migration(apps, schema_editor):
+    """There's no need to do anything special to reverse these migrations."""
+    return
```
Shouldn't it still do something? Can Django already figure out how to, for example, change a field back to True from False on its own?

SebastiaanZ replied:
The reason we don't need to do anything is because we can't reverse the state of the actual data; we simply don't know what the previous value were. Reversal is necessary if the forward function of `RunPython` also changed the model itself, since you then have to restore the model to the state prior to the migration.

So, we can't undo the data migration and we don't have to undo structural changes, since we made none.

MarkKoz replied:
> we simply don't know what the previous value were.

Oh, right... that seems quite obvious now 😄

MarkKoz commented on pydis_site/apps/api/migrations/0044_active_infractions_migration.py:
```diff
@@ -0,0 +1,106 @@
+    )
+
+    # Prefetch their active received infractions of `infr_type` and attach `.infractions` attribute
+    query = query.prefetch_related(
```
Minor efficiency thing: I think the prefetch could be done after the gte 2 filter and the distinct()

MarkKoz replied:
Well it's just a migration I guess this sort of stuff doesn't even matter

SebastiaanZ replied:
It's a simple change, but I wonder if it will make a difference in the SQL query that's going to be made. The actual query will not be run until after all the methods are chained together and we're actually trying to read from `query` in the original function.

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py:
```diff
@@ -0,0 +1,545 @@
+         - `hidden` (default: False).
+
+         The parameters `infraction_model` and `user_model` can be used to pass in an instance of
+         both model classes from a different migration/project state.
```
These two lines seem to be indented a bit too far.

SebastiaanZ replied:
I forgot about this. I still need to fix this.

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py:
```diff
@@ -0,0 +1,545 @@
+
+    def test_infraction_factory_multiple_users(self):
+        """Does the test database hold as many infractions as we tried to create?"""
+        created_infractions = {}
```
This dictionary doesn't seem to be used anywhere else.

SebastiaanZ replied:
I've removed it in my latest commit (not pushed atm, but soon)

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py:
```diff
@@ -0,0 +1,545 @@
+        return InfractionHistory(user_id=user_id, infraction_history=infraction_history)
+
+
+class InfractionFactoryTests(MigrationsTestCase):
```
Why is this a `MigrationsTestCase`? The tests for the factory aren't concerned with what values are allowed for fields but rather just that if a given value is equal to the resulting value. Is it just so future migrations don't break this test (e.g. in the future a check constraint for a kick would be added that it must always be hidden, so line 105 would fail)?

SebastiaanZ replied:
> Is it just so future migrations don't break this test (e.g. in the future a check constraint for a kick would be added that it must always be hidden, so line 105 would fail)?

Yes, it's to pin these tests to the point in the migration history we are testing. For our non-migration tests, we always update the tests to reflect the current (i.e., latest) states of the models, but that would not make sense for something that is used specifically to generate data according to an older model.

In this case, I want to use `InfractionFactory` to specifically generate data that would violate the UniqueTogether constraint introduced in a later migration.

MarkKoz replied:
> In this case, I want to use InfractionFactory to specifically generate data that would violate the UniqueTogether constraint introduced in a later migration.

That makes perfect sense, I did not catch this use myself.

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py:
```diff
@@ -0,0 +1,545 @@
+        )
+
+        # User #1: clean user with no infractions
+        cls.created_infractions.update(
```
Why use `.update()` over `cls.created_infractions[1] = InfractionFactory.create(...)`?

SebastiaanZ replied:
Good point. Adaptation from an earlier version and it was late.

SebastiaanZ replied:
I've changed this.

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py:
```diff
@@ -0,0 +1,545 @@
+        # User #1: clean user with no infractions
+        cls.created_infractions.update(
+            {
+                1: InfractionFactory.create(
```
Do you think using strings as keys instead of IDs would improve readability, or is it already pretty clear from the test case names which infractions are being used?

SebastiaanZ replied:
I think strings like `"user 1"` would be clearer, so I'm going to change that.

MarkKoz replied:
I was thinking something along the lines of `"no_infractions"`, `"one_active_note"`, etc.

MarkKoz replied:
I disliked scrolling back and forth in the module to see what the infractions are supposed to be but I understand the objects need to be created in the hook rather than in the test cases (which would be more ideal for readability).

SebastiaanZ replied:
Yes, that can't really be helped unless we want to rollback, add data, apply the migration in each test case. That would increase the running time of the tests by a lot, though. This test is already on the heavy side (although we're talking one to two seconds in the pipeline, which is acceptable in my opinion).

I'll use sensible strings as keys to make it clearer, I think that's the best we can do here.

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_active_infraction_migration.py:
```diff
@@ -0,0 +1,545 @@
+        user_id, infraction_history = self.created_infractions[2]
+        inactive_note = infraction_history[0]
+        self.assertTrue(
+            self.infraction_model.objects.filter(**model_to_dict(inactive_note)).exists()
```
Do all these filter by infraction ID too? Or is that not necessary?

SebastiaanZ replied:
Yes, they do. The infraction object returned by the InfractionFactory will include the `id` (and all the other fields of the created `Infraction` object).

MarkKoz reviewed (commented):
* Migration testing is quite clever. I'm impressed. Nice work!
* The infractions tests don't strike me as very DRY, though many other tests for the site are probably also guilty of that. Would it be feasible to come up with some functions to make it more DRY, at least for the new infraction tests you've written?
* While your tests are already thorough, I noticed that `MigrationsTestCase` does not itself have any tests. Would it be feasible to come up with tests for that? Is it even worth it?

jchristgit commented on pydis_site/apps/api/migrations/0044_active_infractions_migration.py:
```diff
@@ -0,0 +1,106 @@
+    """Migrates infractions of non-active types to inactive."""
+    inactive_infraction_types = Q(type="note") | Q(type="warning") | Q(type="kick")
+    infraction_model = apps.get_model('api', 'Infraction')
+    infraction_model.objects.filter(inactive_infraction_types).update(active=False)
```
can't we use `type__in` here instead?

```py
.filter(type__in=('note', 'warning', 'kick'))
```

SebastiaanZ replied:
Yes, we can. I'll change that as well.

jchristgit commented on pydis_site/apps/api/migrations/0044_active_infractions_migration.py:
```diff
@@ -0,0 +1,106 @@
+
+    # Get users with active infraction(s) of the provided `infr_type`
+    query = user_model.objects.filter(
+        Q(infractions_received__type=infr_type, infractions_received__active=True)
```
I am probably missing something, but what's the difference between `Q` and no `Q`?

SebastiaanZ replied:
Nothing. I changed my approach, copy-pasted it, and did not think about it properly. I'm working on an update of this PR, so I'll this here as well.

jchristgit commented on pydis_site/apps/api/tests/migrations/base.py:
```diff
@@ -0,0 +1,102 @@
+
+        >>> Infraction = apps.get_model('api', 'Infraction')
+        """
+        pass
```
isn't a docstring without method body valid on its own without `pass`?

MarkKoz replied:
I think so, but I don't mind the `pass` being there to more explicitly convey that it does nothing.

jchristgit replied:
fair

SebastiaanZ committed:
Merge branch 'master' into active-infractions-validation

SebastiaanZ committed:
Solve migration conflict by renaming migrations

The migration files were generated and named before the migrations
added by other pull requests. This caused the migration path to
diverge. Since the migrations did not touch the same models, the
solution was to rename the migration files to place them at the end
of the migration history.

SebastiaanZ committed:
Apply feedback for pull request #278

https://github.com/python-discord/site/pull/278

This commit applies the feedback given to PR #278. I have made a
number of significant changes:

- I have added tests for the `MigrationsTestCase` class, including a
  test to see if the test case travels to the specified points in the
  migration history.

- The dictionary of infracted users in the active migration tests now
  uses informative strings as keys. This makes it easier to relate a
  specific test to the data it's testing.

- I have simplified the migration query logic by replacing unneeded Q
  objects by simpler keyword arguments to `filter` method calls.

- I have removed the custom validation logic for allowing only one
  active infraction of a given type per user and replaced it by a
  `UniqueTogetherValidator` with a custom validation error message.

- `test_unique_constraint_accepts_active_infraction_after_inactive_
   infraction` relied on an Error results instead of a Fail result
   to indicate an issue with the database constraint. Since a fail
   should lead to a Fail status, I've replaced it by a try-except
   block that signals a test failure if an `IntegrityError` was
   caught by it.

MarkKoz commented on pydis_site/apps/api/tests/test_infractions.py:
```diff
@@ -417,27 +424,37 @@ def test_unique_constraint_raises_integrity_error_on_second_active_of_same_type(
+            self.fail("An unexpected IntegrityError was raised.")
+
+    @patch(f"{__name__}.Infraction")
+    def test_the_accepts_active_infraction_after_inactive_infractions_test(self, infraction_patch):
```
This test name is confusing, perhaps unsurprisingly.

MarkKoz replied:
I don't really feel like this test is necessary to be honest.

SebastiaanZ replied:
It's necessary since we include tests in the coverage report and otherwise the `except`-block for the test we're testing here will never be run. 

We may be able to declare an exception to the coverage requirement for those specific lines, but I wrote to test to make sure it actually works like I intended it to, since it does have custom test logic.

MarkKoz replied:
Oh right. I am not used to coverage being enabled for tests. I still think the name could be better by mentioning that it is specifically testing the exception being caught, but the name is already getting quite long 🤔 .

MarkKoz commented on pydis_site/apps/api/tests/migrations/test_base.py:
```diff
@@ -0,0 +1,135 @@
+        self.assertEqual(migration_executor().loader.project_state.call_args_list, expected_args)
+
+    def test_loader_build_graph_gets_called_once(self, migration_executor):
+        """We should rebuild the before applying the second set of migrations."""
```
This docstring seems to be missing a word.

SebastiaanZ replied:
Ah, yes. I'll fix that.

SebastiaanZ committed:
Fix indentation and missing word in docstring

The last two lines in the docstring of `InfractionFactory` in the
`test_active_infraction_migration.py` file were overindented by one
space. I've removed the space.

The docstring of the `test_loader_build_graph_gets_called_once`
method in the `MigrationsTestCaseNoSideEffectsTests` calss of the
`test_base.py` file was missing two words. Corrected.

SebastiaanZ committed:
Make test less fragile and improve test name

The `test_returns_400_for_active_infractions_of_type_that_cannot_be_
active` test relied on the order in which the validation was done
since it contained incompatible combinations of arguments. The test
has been changed to make sure the data is valid except for the thing
we actually want to test.

I have also tried to improve the name of the test that tests the
`test_unique_constraint_accepts_active_infraction_after_inactive_
infraction` test. It now includes the logic of what it does, but
not the entire name of the test it's testing.

SebastiaanZ committed:
Merge branch 'master' into active-infractions-validation

SebastiaanZ committed:
Resolve migration merge conflicts

Since other pull requests were merged that also included migrations
for the API app, this PR needed to be updated to avoid conflicts in
the migration history. In addition, the test files contained names of
specific migration files that needed to be updated to the correct
names after the merge resolution.
